### ISSUE TITLE
Deprecate Flux/Mono#compose 

### ISSUE THREAD

**Issue by smaldini** [reactor/reactor-core#1745]
`transform` and `compose` difference is never obvious when looking at the names only.
`compose` can be achieved with `transform`+`defer`, so if we were to rationalize down to a single method, `transform` would be the 1s candidate.

**Comment by simonbasle** [comment]
@smaldini should we still consider keeping 2 methods, effectively renaming `compose`? eg. `composeLazily`?

**Comment by bsideup** [comment]
@simonbasle or maybe `transformAtSubscribe` / `transformDeferred`?

**Comment by simonbasle** [comment]
After further discussing naming with @smaldini, the `compose` prefix sounds better (we always talk about operator composition, and these are alternative ways to compose preset operators). I feel `composeNow` is explicit enough a replacement for `transform`. For the deferred one, I'm still not sure but candidates include `composeDefer` (the one I went with in my PR, not a verb but kinda references the `defer()` operator), `composeLater` (another strong candidate IMHO), `composeLazily`, `composedDeferred` (probably both a bit too long).

**Comment by bsideup** [comment]
My personal preference: `composeLazily`.

`*Later` and `*Defer` sounds like it will be composed later, but some may think that it will happen only once, just... later.

While "lazy computations" are usually treated as "when the time comes" but also does not put any constraints on how many times it is going to happen

Also, it aligns well with `composeNow`. "Now" vs "Lazily" is easy to explain 👍

**Comment by dfeist** [comment]
I'm probably too late to comment on this now I guess, but as the original contributor of `transform()` I can't say I like the new methods much:

- I do agree `compose` and `transform` can be confusing without javadoc.
- I felt @smaldini's original proposal made the most sense. `transform+defer`
- I feel `transform` makes more sense than `compose` given you are using a transformer function to transform the flux.  (But this depends if you think about what it does, or what it achieves functionally I guess)
- The use of "now" and "later" are not a great way to talk about lifecycle TBH.  Reactor already has the concepts of assembly/subscription/deferring/onNext etc, "now" and "later" are vague and unclear in comparison IMO.  Also `later` implies once, which I don't believe is the case.
- Also, these terms don't match any use in the rest of the API. Why isn't subscribe() subscribeNow()?  Also, shouldn't defer() be renamed to later() to match?

I'm summary:
- I'd ideally stick to the original suggestion, especially if `compose()` isn't very widely used today.
- If you want to keep syntactic sugar for `compose` then think about `transformOnSubscribe`/`transformDeffered` or something similar.
- If renaming transform to compose is a must, then think about using things like `composeOnAssembly` and `composeOnSubscribe` or similar to match existing lifecycle/terms.

**Comment by simonbasle** [comment]
interesting points @dfeist, too bad I just merged the PR today with `composeNow`+`composeLater`, but there is still a tiny amount of time to fix this.

To me `transform` sounded a bit too generic, while we are talking about composing operators via the `Function`, hence my preference for `compose` prefix. But it is just a preference...

I absolutely don't think that `compose()` should entirely go away, even if it isn't widely used.

Your arguments make me rethink the `now` vs `later` prefixes though. The alternatives are all a bit of a mouthful, but eh.

**Comment by dfeist** [comment]
I don't have a major issue with compose instead of transform.  _But_, moving from transform -> compose is a big part of the complication here, else you could just use plain `compose` on not have to have the ugly `Now` suffix.

>don't think that compose() should entirely go away, even if it isn't widely used.
Even if you can achieve the same with `transform+defer`?  I thought there was an overall drive to reduce the number of operators in cases like this.  Do you know how widely it is used?

>Your arguments make me rethink the now vs later prefixes though. 
Yes, these definitely need a rethink! This is the most important part of my feedback.

**Comment by dfeist** [comment]
@simonbasle If there is no good solution, I wouldn't rule out living things as they were as a backstop. Only issues currently is you need to read the doc.

**Comment by simonbasle** [comment]
> I don't have a major issue with compose instead of transform. _But_, moving from transform -> compose is a big part of the complication here, else you could just use plain `compose` on not have to have the ugly `Now` suffix.
> 
right, but not having a suffix at all makes it less clear that there is 2 different path (although sharing the same root might help discovering the alternative API), or might make people think that the `compose()` is the default go-to method and `composeNow()` (or whatever we renamed transform) is for a more specific specialization - like what we have with `concatMap` vs `concatMapDelayError`. Whereas the real choice between the two would be more akin to choosing between `flatMap` and `concatMap` IMO.

That said, maybe the reality of uses cases would align well with the noSuffix/Suffix approach _in the case of `transform()`_, like `transform()` vs `transformOnSubscription()` 🤔 

> > don't think that compose() should entirely go away, even if it isn't widely used.
> 
> Even if you can achieve the same with `transform+defer`? I thought there was an overall drive to reduce the number of operators in cases like this. Do you know how widely it is used?

The dichotomy is important, because people should ask themselves "does this transformation make sense at assembly time or subscription time?". I want to increase the chances people using the API will at least recognizes these two cases, so materializing the two cases in the API is important.

> 
> > Your arguments make me rethink the now vs later prefixes though.
> 
> Yes, these definitely need a rethink! This is the most important part of my feedback, be a shame to have this vagueness/inconsistency in API.

**Comment by simonbasle** [comment]
@smaldini what do you think given @dfeist feedback ? ☝️

**Comment by dfeist** [comment]
>not having a suffix at all makes it less clear that there is 2 different path

Not really, using the same root makes specialized alternatives discoverable in auto-complete.  

A suffix-less default alternative improves usability as you naturally use the option without a suffix until you realize you need something more specialized. On the other hand, if presented with multiple suffixes, the user is forced into a decision one way or other, and this either results in guesswork or them having to go and read docs.  Imagine there being `concatMapEagerError/concatMapDelayError` but not `concatMap`, nightmare for new users.

Anyway.. my suggestion based on i) shared root, ii) default with no suffix, iii) use of existing defer/subscribe vocab/lifecyle is:
```
transform() + 
 - transformDeffered()
 - transformPerSubscribe() 
 - transformOnSubscription() 
```

Could use `compose()` instead, no issue there. Just you have a potential concern with changing the behaviour (but not the name) of `compose`, which you may not see as a major concern, I don't know.

**Comment by simonbasle** [comment]
Thanks for your input @dfeist. Reopening this.

After all these discussions I'm sold on `transform()` + `transformDeferred()`:

 - [x] keeping `transform()` limits the need for a migration for the operator that is most commonly used from what we can gather
 - [x] regroups the operators in autocomplete suggestions
 - [x] `flux.transformDeferred(function)` is still shorter than the ceremony of `Flux.defer(() -> function.apply(flux))`
 - [x] people should be able to link `defer()` and `transformDeferred()`.
 - [x] there's currently no risk of the words defer and deferred to mean something wastly different in Reactor, which wasn't the case for "onSubscribe" and "doOnSubscribe"

**Comment by dfeist** [comment]
@simonbasle Agree on all counts.

> defer and deferred to mean something wastly different in Reactor

Don't quite follow this though. I assume you mean "vastly"?  I see at same vocab:
- `.defer()` == supply publisher *lazily per-subscription*
- `.transformDeferred()` == transform flux *lazily per-subscription*.
